File: backend/kisskh_scraper.py
import requests
import json
import time
from automation import update_react_database, get_movie_data

# Configurations
KISSKH_API_BASE = "https://kisskh.co/api/DramaList/List"
PAGES_TO_SCRAPE = 5 # Each page usually has 30-40 items. Adjust this up to scrape more!
DELAY_BETWEEN_REQUESTS = 1 # Seconds to avoid being blocked by TMDB or Kisskh

def fetch_kisskh_movies(page):
    """Fetch movies/dramas from Kisskh API"""
    params = {
        "page": page,
        "type": 0, # 0 usually means all/mixed, or 1 for movies
        "sub": 0,
        "country": 0,
        "status": 0,
        "order": 1,
        "pageSize": 40
    }
    
    headers = {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36"
    }

    try:
        response = requests.get(KISSKH_API_BASE, params=params, headers=headers, timeout=10)
        if response.status_code == 200:
            return response.json()
        else:
            print(f"Failed to fetch page {page} from KissKH. Status: {response.status_code}")
            return []
    except Exception as e:
        print(f"Error connecting to KissKH: {str(e)}")
        return []

def run_bulk_import():
    print(f"Starting Bulk Import from KissKH (Pages 1 to {PAGES_TO_SCRAPE})...")
    
    total_added = 0
    total_failed = 0
    
    for page in range(1, PAGES_TO_SCRAPE + 1):
        print(f"\n--- Fetching KissKH Page {page} ---")
        items = fetch_kisskh_movies(page)
        
        if not items:
            print("No more items or blocked. Stopping.")
            break
            
        for item in items:
            # KissKH returns title, we need to find it on TMDB to get the High-Quality posters, ID, and details
            title = item.get("title", "")
            if not title:
                continue
                
            print(f"Searching TMDB for: {title}")
            
            # Fetch from TMDB using our existing automation.py function
            tmdb_info = get_movie_data(title)
            
            if tmdb_info:
                print(f"Found TMDB Data! Adding '{tmdb_info['title']}' to React Database...")
                
                # We don't have local video for all of them yet, so we pass None for subtitle
                # The React player will automatically fallback to the streaming Iframe for these!
                update_react_database(
                    movie_info=tmdb_info, 
                    video_url="", # Can be added manually later
                    sub_path=None
                )
                total_added += 1
            else:
                print(f"Warning: Could not find '{title}' on TMDB. Skipped.")
                total_failed += 1
                
            time.sleep(DELAY_BETWEEN_REQUESTS) # Be polite to APIs
            
    print(f"\n=== BULK IMPORT COMPLETE ===")
    print(f"Successfully Added to FlickFlow: {total_added} Movies/Dramas")
    print(f"Failed/Not Found: {total_failed} items")
    print("Refresh your React website (localhost:5173) to see the newly populated AI Processed directory!")

if __name__ == "__main__":
    run_bulk_import()

File: backend/kisskh_scraper.mjs
import fs from 'fs';
import path from 'path';
import https from 'https';

const TMDB_API_KEY = "3fd2be6f0c70a2a598f084ddfb75487c";
const KISSKH_API_BASE = "https://kisskh.co/api/DramaList/List";
const REACT_DATA_PATH = path.resolve("./public/movies.json");

const PAGES_TO_SCRAPE = 10; // Load thousands of Asian Dramas and Movies

const sleep = ms => new Promise(r => setTimeout(r, ms));

async function fetchJSON(url) {
    try {
        const response = await fetch(url, {
            headers: {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
            }
        });
        if (!response.ok) return null;
        return await response.json();
    } catch (e) {
        return null;
    }
}

async function getMovieData(movieName) {
    try {
        const url = `https://api.themoviedb.org/3/search/multi?api_key=${TMDB_API_KEY}&query=${encodeURIComponent(movieName)}`;
        const data = await fetchJSON(url);
        if (data && data.results && data.results.length > 0) {
            return data.results.find(r => r.media_type === 'tv' || r.media_type === 'movie') || data.results[0];
        }
    } catch (e) {
        console.error("TMDB error:", e.message);
    }
    return null;
}

function updateReactDatabase(movieInfo) {
    let data = [];
    try {
        if (fs.existsSync(REACT_DATA_PATH)) {
            data = JSON.parse(fs.readFileSync(REACT_DATA_PATH, 'utf-8'));
        }
    } catch (e) {}

    const newMovie = {
        id: String(movieInfo.id),
        title: movieInfo.title || movieInfo.name,
        overview: movieInfo.overview,
        release_date: movieInfo.release_date || movieInfo.first_air_date,
        poster_path: movieInfo.poster_path || '',
        backdrop_path: movieInfo.backdrop_path || '',
        vote_average: movieInfo.vote_average || 0,
        media_type: movieInfo.media_type || 'tv', // kisskh is usually tv 
        original_language: movieInfo.original_language || 'ko', // kisskh default to korean if missing
        genre_ids: movieInfo.genre_ids || [],
        video_url: '',
        subtitle_file: null,
        is_local: true
    };

    data = data.filter(m => String(m.id) !== String(newMovie.id));
    data.unshift(newMovie);

    fs.writeFileSync(REACT_DATA_PATH, JSON.stringify(data, null, 4));
}

async function fetchKissKHMovies(page) {
    const url = `${KISSKH_API_BASE}?page=${page}&type=0&sub=0&country=0&status=0&order=1&pageSize=40`;
    try {
        return await fetchJSON(url);
    } catch (e) {
        console.error("Failed KissKH fetch:", e.message);
    }
    return [];
}

async function runBulkImport() {
    console.log(`Starting Node.js Bulk Import from KissKH (First ${PAGES_TO_SCRAPE} Pages)...`);
    
    let totalAdded = 0;
    
    for (let page = 1; page <= PAGES_TO_SCRAPE; page++) {
        try {
            console.log(`\n--- Fetching KissKH Page ${page} ---`);
            let items = await fetchKissKHMovies(page);
            
            if (items.data) items = items.data; // sometimes APIs wrap lists in data property
            if (items.list) items = items.list; // or list

            if (!items || !Array.isArray(items) || items.length === 0) {
                console.log("Empty items or failed to fetch. Actually received:", items);
                break;
            }
                
            for (const item of items) {
                try {
                    const title = item.title;
                    if (!title) continue;
                    
                    console.log(`Searching TMDB for: ${title}`);
                    const tmdbInfo = await getMovieData(title);
                    
                    if (tmdbInfo) {
                        const itemName = tmdbInfo.title || tmdbInfo.name;
                        console.log(`Added -> '${itemName}'`);
                        updateReactDatabase(tmdbInfo);
                        totalAdded++;
                    } else {
                        console.log(`Not found -> '${title}'`);
                    }
                } catch (movieErr) {
                    console.error(`Error processing item:`, movieErr.message);
                }
                await sleep(800); // polite delay
            }
        } catch (pageErr) {
            console.error(`CRITICAL: KissKH Page ${page} failed:`, pageErr.message);
        }
    }
    
    console.log(`\n=== BULK IMPORT COMPLETE ===`);
    console.log(`Successfully Added: ${totalAdded} movies to your site!`);
    console.log(`Refresh http://localhost:5173 to see them.`);
}

runBulkImport();
